Core/GDCore/Events/CodeGeneration/EffectsCodeGenerator.cpp

```diff
@@ -15,16 +15,45 @@
 #include "GDCore/Project/Layout.h"
 #include "GDCore/Project/Object.h"
 #include "GDCore/Project/Project.h"
+#include "GDCore/IDE/ProjectBrowserHelper.h"
 
 namespace gd {
 
-void ExposeProjectEffects(
-    const gd::Project& project,
-    const std::function<void(const gd::Effect& effect)>& worker) {
+void EffectsCodeGenerator::DoVisitObject(gd::Object &object) {
+  auto &effects = object.GetEffects();
+  for (std::size_t e = 0; e < effects.GetEffectsCount(); e++) {
+    auto &effect = effects.GetEffect(e);
+    AddEffectIncludeFiles(effect);
+  }
+};
+
+void EffectsCodeGenerator::AddEffectIncludeFiles(const gd::Effect &effect) {
+  // TODO: this browse all the extensions every time we're trying to find
+  // a new effect. Might be a good idea to rework MetadataProvider to be
+  // faster (not sure if it is a bottleneck at all though - but could be
+  // for events code generation).
+  const gd::EffectMetadata &effectMetadata =
+      MetadataProvider::GetEffectMetadata(platform, effect.GetEffectType());
+
+  for (auto &includeFile : effectMetadata.GetIncludeFiles())
+    includeFiles.insert(includeFile);
+};
+
+void EffectsCodeGenerator::GenerateEffectsIncludeFiles(
+    const gd::Platform &platform,
+    gd::Project &project,
+    std::set<gd::String> &includeFiles) {
+  // TODO Add unit tests on this function.
+  
+  // TODO Merge with UsedExtensionsFinder.
+  // Beware that default lights must not include Three.js.

```

Comment:
This is not super clear.

Reply:
Default lights rely on the fact that UsedExtensionsFinder doesn't find extension usage for effects.
We need to make something explicit to avoid future bugs.